spec.bs

```diff
@@ -6,57 +6,940 @@ Status: CG-DRAFT
+
+  [=User agents=] may specify the <dfn for=value>maximum length</dfn> of a [=value struct/value=].
+
+  [=User agents=] may specify a <dfn>default entry lifetime</dfn>, the default [=duration=] between when an [=shared storage database/entry=] is [=shared storage database/store an entry in the database|stored=] and when it expires.
+
+  ## The [=Shared Storage Database|Database=] Algorithms ## {#database-algorithms}
+
+  <div algorithm>
+    To <dfn for="shared storage database">store an entry in the database</dfn>, given a [=/browsing context=] |context|, a [=entry/key=] |key|, and a [=value struct/value=] |value|, run the following steps:
+
+    1. If |key| is not a [=string=] or |key|'s [=string/length=] exceeds the [=key/maximum length=], return false.
+    1. If |value| is not a [=string=] or |value|'s [=string/length=] exceeds the [=value/maximum length=], return false.
+    1. Let |environment| be |context|'s [=active window=]'s [=relevant settings object=].
+    1. [=Enqueue the following steps=] on |context|'s associated [=shared storage database/shared storage database queue=]:
+        1. Let |databaseMap| be the result of running [=obtain a shared storage bottle map=] for |environment|.
+        1. If |databaseMap| is failure, then return null.
+        1. Let |valueStruct| be a new [=entry/value struct=].
+        1. Set |valueStruct|'s [=value struct/value=] to |value|.
+        1. Let |unsafeMoment| be the [=wall clock=]'s [=wall clock/unsafe current time=].
+        1. Let |moment| be the result of running [=coarsen time=] algorithm given |unsafeMoment| and [=wall clock=] as input.
+        1. Let |fromUnixEpochTime| be the [=duration from=] the [=Unix epoch=] to |moment|.
+        1. Set |valueStruct|'s [=value struct/last updated=] to |fromUnixEpochTime|.
+        1. [=map/Set=] |databaseMap|[|key|] to |valueStruct|.
+        1. Return true.
+  </div>
+
+  <div algorithm>
+    To <dfn for="shared storage database">retrieve an entry from the database</dfn>, given a [=/browsing context=] |context| and a [=entry/key=] |key|, run the following steps:
+
+    1. If |key| is not a [=string=] or |key|'s [=string/length=] exceeds the [=key/maximum length=], return null.
+    1. Let |environment| be |context|'s [=active window=]'s [=relevant settings object=].
+    1. [=Enqueue the following steps=] on |context|'s associated [=shared storage database/shared storage database queue=]:
+        1. Let |databaseMap| be the result of running [=obtain a shared storage bottle map=] for |environment|.
+        1. If |databaseMap| is failure, then return null.
+        1. If |databaseMap| does not [=map/contain=] |key|, return undefined.
+        1. Let |valueStruct| be the result of running [=map/Get=] on |databaseMap| with |key|.
+        1. Let |isExpired| be the result of running [=shared storage database/determine whether an entry is expired=] with |valueStruct|.
+        1. If |isExpired| is true, return undefined.
```

Comment:
I think its preferred to combine things like these two steps into a single step.  Instead of creating the `isExpired` variable and only using it in one step, instead write something like: `If the result of running [=shared storage database/determine whether an entry is expired=] given |valueStruct| is true, then return undefined.`

Reply:
Done